### PR TITLE
Changed the ui redirect path from a constant to a flag

### PR DESCRIPTION
**Pull request by konryd**
**What this PR does / why we need it**: It changes the UI redirect path into a environment-settable flag. This makes it possible to serve the /ui requests for cases when the Kubernetes dashboard is replaced by custom app.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -71,6 +71,8 @@ type ServerRunOptions struct {
 
 	MasterCount            int
 	EndpointReconcilerType string
+
+	UIRedirectUrl string
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: s/Url/URL/g

File: pkg/routes/ui.go [hunk 2]
```diff
@@ -22,14 +22,18 @@ import (
 	"k8s.io/apiserver/pkg/server/mux"
 )
 
-const dashboardPath = "/api/v1/namespaces/kube-system/services/kubernetes-dashboard/proxy"
-
 // UIRedirect redirects /ui to the kube-ui proxy path.
-type UIRedirect struct{}
+type UIRedirect struct {
+	DashboardPath string
```
Comment (lavalamp) [inline pkg/routes/ui.go]: `DestinationURL`

File: pkg/routes/ui.go [hunk 3]
```diff
@@ -22,14 +22,18 @@ import (
 	"k8s.io/apiserver/pkg/server/mux"
 )
 
-const dashboardPath = "/api/v1/namespaces/kube-system/services/kubernetes-dashboard/proxy"
-
 // UIRedirect redirects /ui to the kube-ui proxy path.
```
Comment (lavalamp) [inline pkg/routes/ui.go]: `to the given path`

File: pkg/routes/ui.go [hunk 4]
```diff
@@ -22,14 +22,18 @@ import (
...
-type UIRedirect struct{}
+type UIRedirect struct {
+	DashboardPath string
+}
+
+func NewUIRedirect(dashboardPath string) UIRedirect {
```
Comment (lavalamp) [inline pkg/routes/ui.go]: Change the name here, too (it's not necessarily just a path). Add a godoc comment.

File: pkg/routes/ui.go [hunk 5]
```diff
@@ -22,14 +22,18 @@ import (
...
+	return UIRedirect{DashboardPath: dashboardPath}
```
Comment (lavalamp) [inline pkg/routes/ui.go]: are all destinations equally valid?

File: pkg/routes/ui.go (hunk 2 above)
Comment (lavalamp) [reply pkg/routes/ui.go]: Actually `destinationURL` since it doesn't need to be public.

File: pkg/routes/ui.go (hunk 3 above)
Comment (konryd) [reply pkg/routes/ui.go]: Done

**Comment by lavalamp** [comment]
It's a bit late but I'm inclined to take this, since hard coding this path is kinda dumb anyway.

File: pkg/routes/ui.go (hunk 2 above)
Comment (konryd) [reply pkg/routes/ui.go]: Done

**Comment by liggitt** [comment]
I'd rather see the apiserver become less entangled with UIs, not more

File: pkg/routes/ui.go (hunk 4 above)
Comment (konryd) [reply pkg/routes/ui.go]: Done

File: cmd/kube-apiserver/app/options/options.go [hunk 6]
```diff
@@ -188,6 +191,9 @@ func (s *ServerRunOptions) AddFlags(fs *pflag.FlagSet) {
 		"A port range to reserve for services with NodePort visibility. "+
 		"Example: '30000-32767'. Inclusive at both ends of the range.")
 
+	fs.StringVar(&s.UIRedirectUrl, "ui-redirect-url", s.UIRedirectUrl, ""+
+		"URL of the application serving the UI for the cluster.")
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: Suggestion: provide a way to disable this. Setting it to the empty string?

File: pkg/routes/ui.go (hunk 5 above)
Comment (konryd) [reply pkg/routes/ui.go]: Not sure I understand the question. I've manually validated an http redirect to work.

File: pkg/master/master.go [hunk 7]
```diff
@@ -302,7 +303,7 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 	}
 
 	if c.ExtraConfig.EnableUISupport {
```
Comment (lavalamp) [inline pkg/master/master.go]: Is EnableUISupport currently exposed in a flag?

Comment (liggitt) [reply pkg/master/master.go]: no, and access to `/ui` is not allowed as part of standard policy, either

Comment (konryd) [reply pkg/master/master.go]: Didn't see it anywhere.

**Comment by bryk** [comment]
@liggitt 
> I'd rather see the apiserver become less entangled with UIs, not more ("why does my API server have a UI-related flag?"). Do we really want /ui to become an API?

I think it it already an API. There's a lot of blog posts, documentation, links out there that link to /ui. That's why we're having this PR.

**Comment by liggitt** [comment]
> I think it it already an API. There's a lot of blog posts, documentation, links out there that link to /ui. That's why we're having this PR.

What is the contract of `/ui`? Is it required to exist? Is it required to redirect to kube-dashboard, or any UI?

Making the apiserver aware of a UI running against the apiserver seems like a mistake we should be moving away from that, not doubling down on.

**Comment by konryd** [comment]
Should /ui get deprecated, a redirect to the replacement would be a way to introduce users to the change in a non-disruptive manner.

**Comment by liggitt** [comment]
> Should /ui get deprecated, a redirect to the replacement would be a way to introduce users to the change in a non-disruptive manner.

`/ui` is based on assumptions that do not hold for many API servers (see description of https://github.com/kubernetes/kubernetes/pull/53766). Rather than trying to re-introduce it at the cost of re-entangling the apiserver with a UI layer, I'd prefer to remove the incorrect coupling and allow UI's to deploy on top of the API.

**Comment by bryk** [comment]
@liggitt 
I believe the contract for /ui is "it will open Kubernetes UI", though it was never formalized. I'd love to get rid of it, but there is plenty of references to /ui all over internet and even books like "Kubernetes: Up and Running".

Just removing /ui would break flows of beginner users who follow some tutorial. 

I think it'd be good to remove /ui endpoint long term. I don't believe we can do this for 1.9, though.

**Comment by bryk** [comment]
@kubernetes/dashboard-maintainers

**Comment by bryk** [comment]
@liggitt what do you believe would be needed to remove /ui endpoint? In terms of work items.

**Comment by liggitt** [comment]
> I think it'd be good to remove /ui endpoint long term

I agree. Let's not add flags, config, and encourage new use of something we think should be removed.

> @liggitt what do you believe would be needed to remove /ui endpoint? In terms of work items.

1. Announce deprecation
2. Merge https://github.com/kubernetes/kubernetes/pull/53766 after deprecation period

**Comment by konryd** [comment]
@liggitt @bryk @lavalamp 

It looks like there's a general agreement on the need to deprecate the /ui endpoint, but we seem to disagree on the exact way of handling it.

Unless preceded with community outreach, cutting off the /ui will get perceived as abrupt, no matter how long the feature is deprecated. Setting this to a new location will let us move away from it while not surprising users.

**Comment by liggitt** [comment]
> Setting this to a new location will let us move away from it while not surprising users.

Adding config/flags and encouraging new use of it is the opposite of deprecating, and will make it harder to move away from.

**Comment by lavalamp** [comment]
I think I prefer #53766 to this PR. @liggitt can you make sure a note about deprecation ends up in the 1.9 release notes, and put some comments in the code about when it is safe to remove this feature?

(Points that swayed me: /ui isn't safe for https, and a flag is the wrong way to configure something that all apiservers have to be in sync on.)

@konryd I am sorry I don't have a work around for you. You can investigate installing something in the place of the service which does a redirection. I can't promise it will work, though. It is too late in the release process to provide a real solution.

**Pull request closed without merging**